langchain/document_loaders/blockchain.py

```diff
@@ -55,34 +61,77 @@ def __init__(
             raise ValueError(f"Invalid contract address {self.contract_address}")
 
     def load(self) -> List[Document]:
-        url = (
-            f"https://{self.blockchainType}.g.alchemy.com/nft/v2/"
-            f"{self.api_key}/getNFTsForCollection?withMetadata="
-            f"True&contractAddress={self.contract_address}"
-            f"&startToken={self.startToken}"
-        )
+        result = []
+
+        current_start_token = self.startToken
+
+        while True:
+            url = (
+                f"https://{self.blockchainType}.g.alchemy.com/nft/v2/"
+                f"{self.api_key}/getNFTsForCollection?withMetadata="
+                f"True&contractAddress={self.contract_address}"
+                f"&startToken={current_start_token}"
+            )
+
+            response = requests.get(url)
 
-        response = requests.get(url)
+            if response.status_code != 200:
+                raise ValueError(
+                    f"Request failed with status code {response.status_code}"
+                )
 
-        if response.status_code != 200:
-            raise ValueError(f"Request failed with status code {response.status_code}")
+            items = response.json()["nfts"]
 
-        items = response.json()["nfts"]
+            if not items:
+                break
 
-        if not (items):
+            for item in items:
+                content = str(item)
+                tokenId = item["id"]["tokenId"]
+                metadata = {
+                    "source": self.contract_address,
+                    "blockchain": self.blockchainType,
+                    "tokenId": tokenId,
+                }
+                result.append(Document(page_content=content, metadata=metadata))
+
+            if not self.get_all_tokens:
```

Comment:
@skcoirz I appreciate the feedback!  I'm certainly open to suggestions on how to improve the flow of logic here.  The reason I put this self.get_all_tokens check after the first API call (which I agree is maybe a bit counter-intuitive) is because the tokenId can take many formats (e.g. int, hex with leading zeros, etc.) and that format is needed for the 'get_all' call (else it will return an empty array).  So I execute the first API call, and if self.get_all_tokens is true, I determine the tokenId type and use it to paginate through the API.  With either flow, that first call is needed.  If the flag is false, the loop stops after the first call but if true it will continue until no tokens are returned from the API.  

I hope that's clear.  I can certainly change this flow if the logic is unclear.  I just wanted to outline the thinking behind my thought process. Thanks again!

Reply:
I see your point. The first call is shared by both. Even we split the logic after the first call, we still need to make the url call later for get_all_tokens case, which make the code redundant a little bit as the same function is used in 2 sequential places. So like what you did here, a `do, while` makes sense.

I’m thinking of your needs here + our chat of time consumption. Have you considered async calls? We can make the url calls on parallel. I shared my thoughts with Codeium. The following is the sponsored example. :) Please take a look if this makes it more efficient. Thanks!

```python3
import asyncio
import aiohttp
from typing import List
from .document import Document

async def afetch(session: aiohttp.ClientSession, url: str) -> dict:
    async with session.get(url) as response:
        if response.status != 200:
            raise ValueError(f"Request failed with status code {response.status}")
        return await response.json()

async def afetch_all(session: aiohttp.ClientSession, urls: List[str]) -> List[List[dict]]:
    results = await asyncio.gather(*[afetch(session, url) for url in urls])
    return [result["nfts"] for result in results if result["nfts"]]

# TODO: following the convention in this code base, it’s probably better to create another 
# function with name `aload` and handle all the downstream calls in all `a-` prefix 
# functions in the dependent classes.
async def aload(self, max_requests: int = 100) -> List[Document]:
    result = []
    current_start_token = self.startToken
    urls = []
    request_count = 0

    # TODO: I found you don’t have an exit condition check here. AI also failed to detect it. 
    # (very interesting.. clearly it’s a chat bot instead of an analytic bot…) 
    # I’m not really familiar with the use case here, do you have a good suggestion what 
    # can be used as a stricter exit condition? Thanks!
    while request_count < max_requests:
        url = (
            f"https://{self.blockchainType}.g.alchemy.com/nft/v2/"
            f"{self.api_key}/getNFTsForCollection?withMetadata="
            f"True&contractAddress={self.contract_address}"
            f"&startToken={current_start_token}"
        )
        urls.append(url)

        if not self.get_all_tokens:
            break

        current_start_token = self._get_next_tokenId(result[-1].metadata["tokenId"]) if result else self.startToken

        request_count += 1

    async with aiohttp.ClientSession() as session:
        nft_lists = await afetch_all(session, urls)

        for nfts in nft_lists:
            for item in nfts:
                content = str(item)
                tokenId = item["id"]["tokenId"]
                metadata = {
                    "source": self.contract_address,
                    "blockchain": self.blockchainType,
                    "tokenId": tokenId,
                }
                result.append(Document(page_content=content, metadata=metadata))

    if not result:
        raise ValueError(f"No NFTs found for contract address {self.contract_address}")

    return result
```

Reply:
I see your point. The first call is shared by both. Even we split the logic after the first call, we still need to make the url call later for get_all_tokens case, which make the code redundant a little bit as the same function is used in 2 sequential places. So like what you did here, a `do while` makes sense.

Reply:
Sounds good @skcoirz!  Shall I mark this convo as resolved?

Reply:
sounds good!